docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,140 @@
+---
+title: auto-return-agent-to-infra-env
+authors:
+  - "@carbonin"
+creation-date: 2022-05-17
+last-updated: 2022-05-17
+---
+
+# Automatically Return Agents to InfraEnv
+
+## Summary
+
+Without a BareMetalHost there is currently no way to automate the process
+of unbinding a host from a cluster. When a user wants to reuse a host for a new
+cluster they have to use the assisted-service unbind API, then manually boot
+the host back into the discovery image.
+
+To support automating this procedure, the assisted-service will configure nodes
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases where a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+* Allow for a node to be gracefully removed from an installed cluster
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host. This agent will handle
+downloading the required artifacts and rebooting the host into the discovery
+process.
+
+### Timeline
+
+1. User triggers an action which will prepare a host to be unbound
```

Comment:
What happens if the host is a master node? Assisted doesn't know how to add master nodes to a cluster, so it kinda leaves you with a broken cluster that you cannot fix using just assisted

Reply:
I think we can make it as a non goal in the first step (does not mean that we dont have to think about it of course)

Reply:
It's possible that a user would want to reclaim all the nodes in a cluster at once though, right?
I don't think we want to disable this totally for control plane nodes.

I mention this in the "Open Questions" section. Specifically I mention how we target the MachineConfig.

Reply:
Very good point, in case of hypershift it will not be a problem because we don't manage those  and it will mean that user deleted the cluster, we just need to make sure that he is doing it after we unbound the hosts.
In case of full bare metal i think that we can start without it as rom suggested and then improve the process.